libraries/architecture/src/main/kotlin/io/element/android/libraries/architecture/Async.kt

```diff
@@ -18,51 +18,148 @@ package io.element.android.libraries.architecture
 
 import androidx.compose.runtime.MutableState
 import androidx.compose.runtime.Stable
+import kotlin.contracts.ExperimentalContracts
+import kotlin.contracts.InvocationKind
+import kotlin.contracts.contract
 
+/**
+ * Sealed type that allows to model an asynchronous operation.
+ */
 @Stable
 sealed interface Async<out T> {
+
+    /**
+     * Represents a failed operation.
+     *
+     * @param exception the exception that caused the operation to fail.
+     * @param prevData the data returned by a previous successful run of the operation if any.
+     */
+    data class Failure<out T>(
+        val exception: Throwable,
```

Comment:
Since this is not an `Exception` (but a `Throwable`), I would not name it like this.

Maybe use `failure`, keep `error`, or just simple `throwable`. From `mvrx`, it's `error`: [code](https://github.com/airbnb/mavericks/blob/main/mvrx-common/src/main/java/com/airbnb/mvrx/Async.kt#L63).

Reply:
I agree, I took the naming from `Result.kt` as convention. I thought making it consistent with `Result` terminology would be best.
Though I have no strong opinions on this. Happy to improve it.